src/Optimization/hiopDualsUpdater.cpp

```diff
@@ -336,64 +344,53 @@ hiopDualsLsqUpdateLinsysAugSparse::~hiopDualsLsqUpdateLinsysAugSparse()
   delete lin_sys_;
 }
 
-bool hiopDualsLsqUpdateLinsysAugSparse::do_lsq_update(hiopIterate& iter,
-                                                      const hiopVector& grad_f,
-                                                      const hiopMatrix& jac_c,
-                                                      const hiopMatrix& jac_d)
+bool hiopDualsLsqUpdateLinsysAugSparse::
+instantiate_linear_solver(const char* linsol_opt,
```

Comment:
To be honest, I'm thinking if we should move some logics into`hiopOptionsNLP::ensure_consistence()`.
Then we can have a much concise code in here and `hiopKKTLinSysXXX::determineAndCreateLinsys`.
For example, we have already downselect the combination of `KKTLinsys`, `linear_solver_sparse`, `compute_mode` and others in `ensure_consistence()`. Then we can implement `determineAndCreateLinsys` in the base class without worry about host or device, e.g.:
```
  // auto has been replaced by others in the preprocess
  if(linear_solver == "ma57") {
#ifdef HIOP_USE_COINHSL
    nlp_->log->printf( xxxx );
    linSys_ = new hiopLinSolverIndefSparseMA57(n, nnz, nlp_);
#endif // HIOP_USE_COINHSL
  } else  if(linear_solver == "cuslover") {
#ifdef HIOP_USE_CUSOLVER
    nlp_->log->printf( xxxx );
    hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
#endif // HIOP_USE_CUSOLVER
  } else if ...
```

Reply:
I think the proposed change can minimize our future work when some new linear solvers are added.

Reply:
we should definitely update `ensure_consistence` with any new logic . But I think we should keep redundant logic, at least for some time, just to be safe.

Reply:
also, some of the consistency checks for options need to be delayed and done later (in the Alg classes). This occurs because the Alg cannot always be determined based on the options and there are options that depend on/are applicable with specific Alg implementations. See `reload_options` in the Newton IPM Alg class.

Reply:
Got it!